File: litert/kotlin/sample_apps/dummy_sample/src/java/org/tensorflow/tflite/experimental/litert/sample/AlignedBufferUtils.java
package org.tensorflow.tflite.experimental.litert.sample;

import android.annotation.SuppressLint;
import java.lang.reflect.Field;
import java.nio.Buffer;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/**
 * Helper methods for creating ByteBuffers aligned to a 64-byte boundary. This is required if LiteRT
 * is built with LITERT_HOST_MEMORY_BUFFER_ALIGNMENT = 64.
 *
 * <p>WARNING: Reflection on the "address" field can break on certain Android versions or future
 * devices that block reflection. You may need alternative methods.
 */
public final class AlignedBufferUtils {

  private static final int LITERT_ALIGNMENT = 64; // Must match LITERT_HOST_MEMORY_BUFFER_ALIGNMENT

  private AlignedBufferUtils() {
    // Prevent instantiation
  }

  /**
   * Creates a 64-byte aligned ByteBuffer with a capacity of exactly {@code capacityBytes}. The
   * returned ByteBuffer: - is direct - ordered native - starts at a multiple-of-64 address
   */
  public static ByteBuffer create64ByteAlignedByteBuffer(int capacityBytes) {
    // Over-allocate by up to 63 bytes, so we can shift to the next 64 boundary.
    ByteBuffer big =
        ByteBuffer.allocateDirect(capacityBytes + (LITERT_ALIGNMENT - 1))
            .order(ByteOrder.nativeOrder());

    long baseAddress = getAddress(big);
    // Round up baseAddress to next multiple of 64:
    long alignedAddress = (baseAddress + (LITERT_ALIGNMENT - 1)) & ~((long) (LITERT_ALIGNMENT - 1));
    int shift = (int) (alignedAddress - baseAddress);

    // Position 'big' to that shift, then slice so the new ByteBuffer is at that alignment.
    big.position(shift);
    ByteBuffer aligned = big.slice().order(ByteOrder.nativeOrder());
    aligned.limit(capacityBytes); // Ensure exactly capacityBytes

    // Double-check alignment:
    long checkAddr = getAddress(aligned);
    if ((checkAddr % LITERT_ALIGNMENT) != 0) {
      throw new RuntimeException(
          "Failed to achieve 64-byte alignment (address=0x" + Long.toHexString(checkAddr) + ")");
    }
    return aligned;
  }

  /**
   * Reflection-based helper to retrieve the native address of a direct ByteBuffer. This may fail if
   * hidden APIs or reflection are blocked on your Android platform.
   */
  private static long getAddress(ByteBuffer buffer) {
    try {
      @SuppressLint("DiscouragedPrivateApi")
      Field addressField = Buffer.class.getDeclaredField("address");
      addressField.setAccessible(true);
      return addressField.getLong(buffer);
    } catch (Exception e) {
      throw new RuntimeException("Could not get native address via reflection", e);
    }
  }
}

File: litert/kotlin/src/main/kotlin/com/google/ai/edge/litert/AlignedBufferUtils.kt
package com.google.ai.edge.litert

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Utility for creating ByteBuffers aligned to a 64-byte boundary.
 *
 * This class uses JNI to determine the actual memory address of direct ByteBuffers
 * and creates properly aligned buffers for optimal performance with LiteRT operations.
 */
object AlignedBufferUtils {
  /** Required memory alignment for LiteRT operations in bytes. */
  private const val LITERT_ALIGNMENT = 64

  init {
    // Load the native library containing the JNI implementation
    System.loadLibrary("litert_jni")
  }

  /**
   * Creates a ByteBuffer aligned to a 64-byte boundary.
   *
   * @param capacityBytes The desired capacity of the buffer in bytes.
   * @return A direct ByteBuffer with the requested capacity, aligned to a 64-byte boundary.
   * @throws RuntimeException If alignment cannot be achieved.
   */
  @JvmStatic
  fun create64ByteAlignedByteBuffer(capacityBytes: Int): ByteBuffer {
    // Allocate extra space to ensure we can achieve alignment
    val buffer = ByteBuffer.allocateDirect(capacityBytes + LITERT_ALIGNMENT - 1)
      .order(ByteOrder.nativeOrder())

    // Get the actual memory address of the buffer
    val baseAddress = nativeGetDirectBufferAddress(buffer)
    if (baseAddress == 0L) {
      throw RuntimeException("Failed to get direct buffer address for alignment")
    }

    // Calculate the aligned address by rounding up to the next multiple of LITERT_ALIGNMENT
    val alignedAddress = (baseAddress + (LITERT_ALIGNMENT - 1)) and 
      ((LITERT_ALIGNMENT - 1).inv().toLong())
    val shift = (alignedAddress - baseAddress).toInt()

    // Create a slice of the original buffer at the aligned position
    buffer.position(shift)
    val alignedBuffer = buffer.slice().order(ByteOrder.nativeOrder())
    alignedBuffer.limit(capacityBytes)

    // Verify the alignment was successful
    val checkAddr = nativeGetDirectBufferAddress(alignedBuffer)
    if (checkAddr == 0L || (checkAddr % LITERT_ALIGNMENT) != 0L) {
      throw RuntimeException("Failed to achieve 64-byte alignment (address=$checkAddr)")
    }
    
    return alignedBuffer
  }

  /**
   * Returns the memory address of a direct ByteBuffer.
   *
   * @param buffer The direct ByteBuffer to get the address for.
   * @return The memory address as a long, or 0 if the operation failed.
   */
  @JvmStatic private external fun nativeGetDirectBufferAddress(buffer: ByteBuffer): Long
}
